### PR TITLE
SDK - Components - Added ComponentStore search

### PR DESCRIPTION
**Pull request by Ark-kun**
ComponentStore(...).search searches for components by name in the configured component store. It prints the name and URL for components that match the given name.
Only components on GitHub are currently supported.

Example:

```
kfp.components.ComponentStore.default_store.search('xgboost')

>>> Xgboost train   https://raw.githubusercontent.com/.../components/XGBoost/Train/component.yaml
>>> Xgboost predict https://raw.githubusercontent.com/.../components/XGBoost/Predict/component.yaml
```

The `default_store` is configured with the main KFP component repo. The users can customize the store with their own repos and URL prefixes:

```
my_store = ComponentStore(
    url_search_prefixes=[
        'https://raw.githubusercontent.com/Ark-kun/kfp_components/master/components/',
        'https://raw.githubusercontent.com/<org>/<repo>/master/components/',
    ],
)
my_store.search('xgboost')
```

### CODE REVIEW

File: sdk/python/kfp/components/_component_store.py [hunk 1]
```diff
@@ -129,6 +142,106 @@ def _load_component_from_ref(self, component_ref: ComponentReference) -> Callabl
...
+                    blob_hash = candidate['git_blob_hash']
+                    if not self._git_blob_hash_to_data_db.exists(blob_hash):
+                        logging.debug('Downloading component spec from "{}"'.format(component_url))
+                        response = requests.get(component_url)
+                        response.raise_for_status()
+                        component_data = response.content
+
+                        # Verifying the hash
+                        received_data_hash = _calculate_git_blob_hash(component_data)
+                        assert received_data_hash.lower() == blob_hash.lower()
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply sdk/python/kfp/components/_component_store.py:191]: Added a proper exception. But generally the user can do nothing here: It's either network is broken or GitHub is corrupted.

File: sdk/python/kfp/components/_component_store.py [hunk 2]
```diff
@@ -129,6 +142,106 @@ def _load_component_from_ref(self, component_ref: ComponentReference) -> Callabl
...
+
+
+def _calculate_component_digest(data: bytes) -> str:
+    return hashlib.sha256(data.replace(b'\r\n', b'\n')).hexdigest()
+
+
+def _list_candidate_component_uris_from_github_repo(url_search_prefix: str) -> Iterable[str]:
+    (schema, _, host, org, repo, ref, path_prefix) = url_search_prefix.split('/', 6)
+    for page in range(1, 999):
+        search_url = 'https://api.github.com/search/code?q=filename:{}+repo:{}/{}&page={}&per_page=1000'.format(_COMPONENT_FILENAME, org, repo, page)
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply sdk/python/kfp/components/_component_store.py:223]: Fixed.

File: sdk/python/kfp/components/_component_store.py [hunk 3]
```diff
@@ -129,6 +142,106 @@ def _load_component_from_ref(self, component_ref: ComponentReference) -> Callabl
...
+        response.raise_for_status()
+        result = response.json()
+        items = result['items']
+        if not items:
+            break
+        for item in items:
+            html_url = item['html_url']
+            # Constructing direct content URL
+            # There is an API (/repos/:owner/:repo/git/blobs/:file_sha) for getting the blob content, but it requires decoding the content.
+            raw_url = html_url.replace('https://github.com/', 'https://raw.githubusercontent.com/').replace('/blob/', '/', 1)
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply sdk/python/kfp/components/_component_store.py:234]: Fixed.

File: sdk/python/kfp/components/_component_store.py [hunk 4]
```diff
@@ -129,6 +142,106 @@ def _load_component_from_ref(self, component_ref: ComponentReference) -> Callabl
...
+                    component_url = candidate['url']
+                    if self._url_to_info_db.exists(component_url):
+                        continue
+
+                    logging.debug('Found new component URL: "{}"'.format(component_url))
+
+                    blob_hash = candidate['git_blob_hash']
+                    if not self._git_blob_hash_to_data_db.exists(blob_hash):
+                        logging.debug('Downloading component spec from "{}"'.format(component_url))
+                        response = requests.get(component_url)
```
[the remark this answers is not in the document]
Comment (Ark-kun) [reply sdk/python/kfp/components/_component_store.py]: Added proper retries here and in another place.

**Comment by Ark-kun** [comment]
> Is that possible to add a test of this library, for example, by either mocking requests.get() or test against current KFP repo?

It would be great to have. I'm experimenting with mocks and plan to add mock-based tests for `ComponentStore` in a future PR. Do you think it's OK?

**Comment by Bobgy** [comment]
This is a great idea!
/lgtm
I'd like to understand more CUJs, like how third parties can share their store how to discover other stores?

**Comment by Ark-kun** [comment]
> I'd like to understand more CUJs, like how third parties can share their store how to discover other stores?

Currently the search only occurs in the configured repositories. Currently there are some issues with global search (it requires API token and it can only list 1000 results). The repo marker file will make the discovery more reliable, but only when users start adopting it.

The `ComponentStore.default_store` is configured with the main KFP component repo. The users can customize the store with their own repos and URL prefixes:

```
my_store = ComponentStore(
    url_search_prefixes=[
        'https://raw.githubusercontent.com/kubeflow/pipelines/master/components/',
        'https://raw.githubusercontent.com/Ark-kun/kfp_components/master/components/',
        'https://raw.githubusercontent.com/<org>/<repo>/master/components/',
    ],
)
my_store.search('xgboost')
```

**Comment by numerology** [comment]
>It would be great to have. I'm experimenting with mocks and plan to add mock-based tests for ComponentStore in a future PR. Do you think it's OK?

Sure. SGTM

**Pull request merged**